lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -1,139 +1,158 @@
-struct YieldNow {
-    yielded: bool,
-}
+    /// Block on the result of an `async` computation from an instance run by `Instance::run_async`.
+    ///
+    /// The primary reason you may want to use `try_block_on` manually is to provide a fallback
+    /// implementation for if your hostcall is called from outside of an asynchronous context.
+    ///
+    /// If `Instance::run_async` is not being used to run the VM, this function will return
+    /// `Err(BlockOnError::NeedsAsyncContext)`.
+    #[doc(hidden)]
+    pub fn try_block_on<R>(&self, mut f: impl Future<Output = R>) -> Result<R, BlockOnError> {
+        // We pin the future to the stack (a requirement for being able to poll the future).
+        // By pinning to the stack instead of using `Box::pin`, we avoid heap allocations for immediately-ready futures.
+        //
+        // SAFETY: We must uphold the invariants of Pin, namely that future does not move until it is dropped.
+        // By overriding the variable named `f`, it is impossible to access f again, except through the pinned reference.
+        let mut f = unsafe { Pin::new_unchecked(&mut f) };
 
-impl YieldNow {
-    fn new() -> Self {
-        Self { yielded: false }
-    }
-}
+        loop {
+            // Get the AsyncContext, or die if we aren't async
+            let arc_cx = match &self.instance().state {
+                State::Running {
+                    async_context: Some(cx),
+                } => cx.clone(),
+                State::Running {
+                    async_context: None,
+                } => return Err(BlockOnError::NeedsAsyncContext),
+                _ => unreachable!("Access to vmctx implies instance is Running"),
+            };
+
+            // build an std::task::Context
+            let mut cx = Context::from_waker(&arc_cx.waker);
+
+            match f.as_mut().poll(&mut cx) {
+                Poll::Ready(ret) => return Ok(ret),
+                Poll::Pending => {
+                    // The future is pending, so we need to yield to the async executor
+                    self.yield_impl::<AsyncYielded, AsyncResume>(AsyncYielded, false, false);
+
+                    // Check that the async context hasn't changed (this could happen if the instance yielded)
+                    match &self.instance().state {
+                        State::Running { async_context: Some(new_cx) } => {
+                            let same_waker = Arc::ptr_eq(&arc_cx, &new_cx) || arc_cx.waker.will_wake(&new_cx.waker);
+
+                            if !same_waker {
+                                // The AsyncContext changed on us. This is because the instance is running from a new RunAsync.
+                                // This probably happened because the instance yielded and was resumed up by `resume_async`.
+                                //
+                                // Poll the future again before yielding to the executor in order to register the new waker.
```

Comment:
The safety here is somewhat difficult to reason about for me. Is there a race where the underlying future still has the old waker, we've been resumed with the new one, and we might miss a wakeup? If not, why not (let's include the reasoning here)?

Reply:
Yes, that issue exists. poll is synchronous, so by the time we have reached this point, it's certain if we are executing with a new AsyncContext, and possible that the future will awake (or has awoken) the old waker, which will mean we miss a wakeup.

Rather than try to prevent this from happening at all and writing our own waker (which is, bluntly: not fun, you basically have to manually implement Arc<dyn Fn() -> ()> with a custom vtable), we check if the waker changed from under us, and if so, we simply poll the future again, which registers the new waker, based on the API contract for Future::poll. If the future is Ready, there's no need for another wakeup, because we can make progress without yielding.

Also: Waking a waker unnecessarily will not cause unsafety or logic errors. In the worst case, the async executor may waste CPU time polling pending futures an extra time. It's also valid to hold onto a waker nearly indefinitely (again, they are roughly equivalent to Arc<dyn Fn()>, just with a manual vtable).

I'm realizing that I don't have a test that would cause this branch to execute. I'll write one. Luckily, it's easy behavior to force to happen (simply switch the async runtime between yield points)

Reply:
OK, so I think I've convinced myself that this is sound. The potentially problematic case that I was considering earlier went something like: new executor polls our outer future; concurrently, underlying async IO or whatever completes, and the inner future we've wrapped notifies its (old) waker; then, we poll future to "reset to new waker". So we indeed do not set the waker in time, but we've also just polled the future, which is exactly what it wanted anyway. Either the underlying wakeup interleaves before our poll (in which case our poll just happens to be correct) or after our poll (in which case it'll pick up the new waker and trigger another wakeup later).

Could you add a comment here so that a later reader also confused about this gets it more easily? :-)